server/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -403,25 +409,78 @@ private static boolean isContentTypeDisallowed(RestRequest request) {
             && SAFELISTED_MEDIA_TYPES.contains(request.getParsedContentType().mediaTypeWithoutParameters());
     }
 
-    private boolean handleNoHandlerFound(String rawPath, RestRequest.Method method, String uri, RestChannel channel) {
+    private boolean handleNoHandlerFound(
+        ThreadContext threadContext,
+        String rawPath,
+        RestRequest.Method method,
+        String uri,
+        RestChannel channel
+    ) {
         // Get the map of matching handlers for a request, for the full set of HTTP methods.
         final Set<RestRequest.Method> validMethodSet = getValidHandlerMethodSet(rawPath);
         if (validMethodSet.contains(method) == false) {
             if (method == RestRequest.Method.OPTIONS) {
+                startTrace(threadContext, channel);
                 handleOptionsRequest(channel, validMethodSet);
                 return true;
             }
             if (validMethodSet.isEmpty() == false) {
                 // If an alternative handler for an explicit path is registered to a
                 // different HTTP method than the one supplied - return a 405 Method
                 // Not Allowed error.
+                startTrace(threadContext, channel);
                 handleUnsupportedHttpMethod(uri, method, channel, validMethodSet, null);
                 return true;
             }
         }
         return false;
     }
 
+    private void startTrace(ThreadContext threadContext, RestChannel channel) {
+        startTrace(threadContext, channel, null);
+    }
+
+    private void startTrace(ThreadContext threadContext, RestChannel channel, String restPath) {
+        final RestRequest req = channel.request();
+        if (restPath == null) {
+            restPath = req.path();
+        }
+        String method = null;
+        try {
+            method = req.method().name();
+        } catch (IllegalArgumentException e) {
+            // Invalid methods throw an exception
+        }
+        String name;
+        if (method != null) {
+            name = method + " " + restPath;
+        } else {
+            name = restPath;
+        }
+
+        Map<String, Object> attributes = new HashMap<>();
```

Comment:
Maybe pre-size this correctly e.g. using `Maps.` utilities? This isn't on the super hot path but sill it's on a network thread :)